sw/host/opentitanlib/src/image/image.rs

```diff
@@ -102,120 +104,110 @@ impl Image {
+        let ext_bytes = entry.to_vec();
+        let end_index = offset
+            .checked_add(ext_bytes.len())
             .ok_or(ImageError::ExtensionOverflow)?;
         let extension_slice = self
             .data
             .bytes
-            .get_mut(self.size..end_index)
+            .get_mut(offset..end_index)
             .ok_or(ImageError::ExtensionOverflow)?;
-        extension_slice.copy_from_slice(bytes);
-        self.size += bytes.len();
+        extension_slice.copy_from_slice(ext_bytes.as_slice());
+        self.size = std::cmp::max(end_index, self.size);
+
+        let mut manifest = self.borrow_manifest_mut()?;
+        manifest.extensions.entries = ext_table;
+
         Ok(())
     }
 
-    /// Sets the extension entry at the given index.
-    pub fn set_extension_entry(
-        &mut self,
-        index: usize,
-        identifier: u32,
-        offset: u32,
-    ) -> Result<()> {
-        let manifest = self.borrow_manifest_mut()?;
-        *manifest
+    /// Allocates space for a manifest extension and sets the offset in the extension table.
+    ///
+    /// This function is similar to `add_manifest_extension`, but doesn't populate any data for the
+    /// extension. This is necessary to properly set the `length` field of the manifest and the
+    /// `offset` field of the manifest extension entry for signing.
+    pub fn allocate_manifest_extension(&mut self, id: u32, len: usize) -> Result<()> {
+        let offset = self.size as u32;
+        self.borrow_manifest_mut()?
             .extensions
             .entries
-            .get_mut(index)
-            .ok_or(ImageError::BadExtensionTableIndex)? =
-            ManifestExtTableEntry { identifier, offset };
-        Ok(())
-    }
+            .iter_mut()
+            .find(|e| e.identifier == id)
+            .ok_or(ImageError::NoExtensionTableEntry(id))?
+            .offset = offset;
 
-    /// Adds an extension to the signed region of this `Image`.
-    pub fn add_signed_manifest_extensions(&mut self, spec: ManifestExtSpec) -> Result<()> {
-        self.add_manifest_extensions(&spec.signed_region, spec.source_path())
-    }
+        self.size = self
+            .size
+            .checked_add(len)
+            .ok_or(ImageError::ExtensionOverflow)?;
 
-    /// Adds an extension to the unsigned region of this `Image`.
-    pub fn add_unsigned_manifest_extensions(&mut self, spec: ManifestExtSpec) -> Result<()> {
-        self.add_manifest_extensions(&spec.unsigned_region, spec.source_path())
+        Ok(())
     }
 
-    fn add_manifest_extensions(
-        &mut self,
-        entries: &[ManifestExtEntrySpec],
-        relative_path: Option<&Path>,
-    ) -> Result<()> {
+    /// Clears any entry in the manifest extension table that doesn't have an offset.
+    ///
+    /// This allows a manifest definition with a populated extension table to be used even when
+    /// extensions aren't provided.
+    pub fn drop_null_extensions(&mut self) -> Result<()> {
```

Comment:
nit: it's not clear to me why we need this -- is there a reason for setting the type of an entry in the extension table but not setting the offset?

Reply:
They are set by `manifest.hjson` where you made your earlier comment https://github.com/lowRISC/opentitan/pull/19134#discussion_r1258364692. Without this function the entries would stick around in the `extensions` table of the manifest with a 0 `offset`.